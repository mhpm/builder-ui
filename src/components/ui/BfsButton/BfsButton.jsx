/** @jsxImportSource @emotion/react */ 
import { memo, useCallback } from 'react';
import PropTypes from 'prop-types';
import { css } from '@emotion/react';
import { BfsIcon } from '../BfsIcon';

export const modes = {
  primary: 'primary',
  secondary: 'secondary',
  negative: 'negative',
  positive: 'positive',
  primaryGhost: 'primary-ghost',
  secondaryGhost: 'secondary-ghost',
  negativeGhost: 'negative-ghost',
  positiveGhost: 'positive-ghost',
};

const modeStyle = (mode) => css`
  background-color: var(--rol-${mode}-bg-color);
  border-color: var(--rol-${mode}-border-color);
  border-width: var(--rol-${mode}-border-thick);
  color: var(--rol-${mode}-text-color);
  font-size: var(--rol-${mode}-text-size);
  font-weight: var(--rol-${mode}-text-weight);

  &:hover,
  &:focus {
    background-color: var(--rol-${mode}-bg-hover-color);
    border-color: var(--rol-${mode}-border-hover-color);
    ${mode.includes('ghost') && `color: var(--color-white)`};
  }

  &:active {
    background-color: var(--rol-${mode}-bg-active-color);
    border-color: var(--rol-${mode}-border-active-color);
  }

  &[disabled] {
    background-color: var(--rol-${mode}-bg-disabled-color);
    border-color: var(--rol-${mode}-border-disabled-color);
    ${mode.includes('ghost') && `color: var(--rol-${mode}-border-color)`};
    cursor: not-allowed;
  }
`;

const stBfsButton = css`
  border: solid;
  border-radius: var(--rol-border-radius);
  column-gap: var(--sm-6xs, .25rem);
  cursor: pointer;
  margin: var(--bfs-button-margin, var(--sp-5xs, 5px));
  min-height: 30px;
  min-width: 70px;
  outline: 0;
  padding: 9px var(--sp-xs, 12px) var(--sp-3xs, 8px) var(--sp-xs, 12px);
  text-transform: uppercase;
`;

const stContainer = css`
  align-items: center;
  display: inline-flex;
  justify-content: center;
`;

const stFluid = css`
  margin: var(--bfs-button-margin, var(--sp-5xs, 5px) 0);
  width: 100%;
`;

const stLoading = css`
  cursor: wait;

  & .content {
    opacity: 0;
  }
`;

const stModes = {
  primary: modeStyle(modes.primary),
  secondary: modeStyle(modes.secondary),
  negative: modeStyle(modes.negative),
  positive: modeStyle(modes.positive),
  primaryGhost: modeStyle(modes.primaryGhost),
  secondaryGhost: modeStyle(modes.secondaryGhost),
  negativeGhost: modeStyle(modes.negativeGhost),
  positiveGhost: modeStyle(modes.positiveGhost),
  tableHeader: css`
    border: none;
    cursor: pointer;
    font-weight: bold;
  `,
};

const stIcon = (loading) => css`
  align-items: center;
  display: ${loading ? 'flex' : 'none'};
  justify-content: center;
  position: absolute;
`;

/**
 * General and reusable button for all the projects, that follows the color
 * conventions from the BFS Design and Component Library. It can be set as
 * `fluid`, `loading`, and `disabled` modes using those configuration flags
 * properties.
 *
 * It comes in 4 different color schemas to chose using the `mode` property:
 * ***primary***, ***secondary***, ***positive***, and ***negative*** each of
 * then in two versions, *regular* and *ghost*.
 *
 * The button label can be set by the `label` property or if this property is
 * not set, by its `children` content. If there are no children either, it will
 * be set to ***BUTTON***.
 *
 * It will receive at the `onClick` property, a *callback* function handler to
 * run when the button will be clicked. That callback will return the value the
 * property `param` was initialized with if it was set.
 *
 * ### Styling
 * `BfsButton` provides the following custom properties for styling:
 *
 * | Custom property      | Description                | Default              |
 * |----------------------|----------------------------|---------------------:|
 * | --bfs-button-margin  | Button margin              | var(--sp-5xs, 5px)   |
 *
 */
export const BfsButton = memo(
  ({
    'data-testid': dataTestId,
    loading,
    icon,
    disabled,
    fluid,
    label,
    mode,
    onClick,
    param,
    escEnable,
    children,
  }) => {
    const clickHandler = useCallback((ev) => {
      ev.stopPropagation();
      ev.preventDefault();

      if (onClick && !loading && !disabled) {
        if (param !== null) {
          onClick(param);
        } else {
          onClick();
        }
      }
    }, [onClick, param, loading, disabled]);

    const keyDownHandler = useCallback((ev) => {
      ev.stopPropagation();
      if (ev.code === 'Space' || ev.code === 'Enter' || (escEnable && ev.code === 'Escape')) {
        clickHandler(ev);
      }
    }, [clickHandler, escEnable]);

    const content = children || label || 'button';

    return (
      <button
        type="button"
        data-testid={dataTestId}
        css={[
          mode !== 'tableHeader' && stBfsButton,
          loading && stLoading,
          stModes[mode],
          fluid && stFluid,
          stContainer,
        ]}
        onClick={clickHandler}
        onKeyDown={keyDownHandler}
        disabled={disabled}
      >
        <>
          <span className="content">
            {content}
          </span>
          <span css={stIcon(loading)}>
            <BfsIcon data-testid={`icon-${icon}`} icon={icon} spin />
          </span>
        </>
      </button>
    );
  },
);

BfsButton.displayName = 'BfsButton';

BfsButton.propTypes = {
  /**
   * Test id only for unit testing element selection.
   */
  'data-testid': PropTypes.string,

  /**
   * This flag set the button in loading mode. When in loading mode a
   * customizable spinner appears as content for the button.
   */
  loading: PropTypes.bool,

  /**
   * The name of the icon to show as spinner when the button is in loading
   * mode. The icon will being shown instead of the text string from `label`
   * property.
   */
  icon: PropTypes.oneOf(['spinner', 'cog', 'circlenotch']),

  /**
   * This flag set the button disabled. When is disabled, the button click event
   * it's not available and the styles change to reflect the unavailability of
   * the button.
   */
  disabled: PropTypes.bool,

  /**
   * This flag set the button in fluid mode. When is fluid, the button will grow
   * in width to use all the available space. If more than one button are
   * together in fluid mode and their parent container has it's `display`
   * property to `flex`, then their will share the available free space.
   */
  fluid: PropTypes.bool,

  /**
   * This is the text string to show as button label.
   */
  label: PropTypes.string,

  /**
   * The button mode that define the selected color scheme to use. It comes in 4
   * different color schemas, each of then in two versions, *regular* and
   * *ghost*.
   */
  mode: PropTypes.oneOf([
    'primary',
    'secondary',
    'positive',
    'negative',
    'primaryGhost',
    'secondaryGhost',
    'positiveGhost',
    'negativeGhost',
    'tableHeader',
  ]),

  /**
   * Callback function to run when the button is clicked. It will return the
   * value the `param` property was set with, if set.
   */
  onClick: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),

  /**
   * Param to be return with the callback function call set by `onClick`
   * property, if anything was set.
   */
  param: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.array,
    PropTypes.string,
    PropTypes.object,
    PropTypes.bool,
    PropTypes.number,
  ]),

  /**
   * Flag that enable the callback execution pressing the "Escape" key.
   */
  escEnable: PropTypes.bool,
};

BfsButton.defaultProps = {
  'data-testid': undefined,
  loading: false,
  icon: 'spinner',
  disabled: false,
  fluid: false,
  mode: 'primary',
  onClick: null,
  param: null,
  escEnable: false,
  label: 'button',
};
