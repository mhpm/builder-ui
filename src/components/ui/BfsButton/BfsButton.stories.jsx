/* eslint-disable react/jsx-props-no-spreading */
import { action } from '@storybook/addon-actions';

import { BfsButton } from './BfsButton';

export default {
  title: 'Atoms/BfsButton',
  component: BfsButton,
  parameters: {
    componentSubtitle: 'Reusable General Button for BFS projects',
  },
  argTypes: {
    'data-testid': {
      control: 'none',
    },
    icon: {
      control: {
        type: 'select',
        options: [
          'spinner',
          'cog',
          'circle',
        ],
      },
    },
    onClick: {
      control: 'none',
    },
    param: {
      control: 'none',
    },
  },
  args: {
    label: 'Button',
    mode: 'primary',
    disabled: false,
    fluid: false,
    loading: false,
    icon: 'spinner',
    onClick: action('Button Clicked'),
    escEnable: false,
  },
  excludeStories: /.*Data$/,
};

export const actionsData = {
  btn1Handle: action('onBtn1'),
  btn2Handle: action('onBtn2'),
  btn3Handle: action('onBtn3'),
  btn4Handle: action('onBtn4'),
  btn5Handle: action('onBtn5'),
  btn6Handle: action('onBtn6'),
  btn7Handle: action('onBtn7'),
  btn8Handle: action('onBtn8'),
};

export const Default = {};

export const Children = {
  render: (args) => (
    <div>
      <BfsButton {...args} onClick={actionsData.btn1Handle}>
        Loren ipsum
      </BfsButton>

      <BfsButton {...args} onClick={actionsData.btn2Handle}>
        <h1>Howdy y&apos;all!</h1>
      </BfsButton>

      <BfsButton {...args} onClick={actionsData.btn3Handle}>
        <h3>Howdy y&apos;all!</h3>
      </BfsButton>
    </div>
  ),
};

export const Loading = {
  args: {
    loading: true,
  },
  render: (args) => (
    <div>
      <BfsButton {...args} onClick={actionsData.btn1Handle} />
      <BfsButton
        {...args}
        mode="secondary"
        onClick={actionsData.btn2Handle}
      />
      <BfsButton
        {...args}
        mode="positive"
        onClick={actionsData.btn3Handle}
      />
      <BfsButton
        {...args}
        mode="negative"
        onClick={actionsData.btn4Handle}
      />
      <BfsButton
        {...args}
        mode="negative"
        onClick={actionsData.btn4Handle}
        label="Button with different size"
      />
      <BfsButton
        {...args}
        mode="negative"
        onClick={actionsData.btn4Handle}
      >
        <h2>Button</h2>
      </BfsButton>
    </div>
  ),
};

export const Fluid = {
  args: {
    fluid: true,
  },
  render: (args) => <BfsButton {...args} />,
};

export const FluidMany = {
  args: {
    fluid: true,
  },
  render: (args) => (
    <div>
      <h3>
        The parent container elment should have
        <em>&apos;display: flex&apos;</em>
      </h3>
      <div style={{ display: 'flex', border: '1px solid #ee00ee' }}>
        <BfsButton
          {...args}
          mode="negative"
          label="delete"
          onClick={actionsData.btn1Handle}
        />
        <BfsButton
          {...args}
          mode="secondary"
          label="cancel"
          onClick={actionsData.btn2Handle}
        />
      </div>
    </div>
  ),
};

export const GhostMode = {
  render: (args) => (
    <div>
      <BfsButton {...args} mode="primaryGhost" />
      <BfsButton {...args} mode="secondaryGhost" />
      <BfsButton {...args} mode="negativeGhost" />
      <BfsButton {...args} mode="positiveGhost" />
    </div>
  ),
};

export const tableHeader = {
  render: (args) => (
    <div>
      <BfsButton {...args} mode="tableHeader" label="Header" />
      <BfsButton {...args} mode="tableHeader" label="Header" />
      <BfsButton {...args} mode="tableHeader" label="Header" />
      <BfsButton {...args} mode="tableHeader" label="Header" />
    </div>
  ),
};

export const Gallery = {
  args: {
    param: 'params',
  },
  render: (args) => (
    <div>
      <div>
        <BfsButton
          {...args}
          mode="primary"
          label="primary"
          onClick={actionsData.btn1Handle}
        />
        <BfsButton
          {...args}
          mode="secondary"
          label="secondary"
          onClick={actionsData.btn2Handle}
        />
        <BfsButton
          {...args}
          mode="negative"
          label="negative"
          onClick={actionsData.btn3Handle}
        />
        <BfsButton
          {...args}
          mode="positive"
          label="positive"
          onClick={actionsData.btn4Handle}
        />
      </div>
      <div>
        <BfsButton
          {...args}
          mode="primary"
          label="primary"
          onClick={actionsData.btn1Handle}
          disabled
        />
        <BfsButton
          {...args}
          mode="secondary"
          label="secondary"
          onClick={actionsData.btn2Handle}
          disabled
        />
        <BfsButton
          {...args}
          mode="negative"
          label="negative"
          onClick={actionsData.btn3Handle}
          disabled
        />
        <BfsButton
          {...args}
          mode="positive"
          label="positive"
          onClick={actionsData.btn4Handle}
          disabled
        />
      </div>
      <div>
        <BfsButton
          {...args}
          mode="primary"
          label="primary"
          onClick={actionsData.btn1Handle}
          loading
        />
        <BfsButton
          {...args}
          mode="secondary"
          label="secondary"
          onClick={actionsData.btn2Handle}
          loading
        />
        <BfsButton
          {...args}
          mode="negative"
          label="negative"
          onClick={actionsData.btn3Handle}
          loading
        />
        <BfsButton
          {...args}
          mode="positive"
          label="positive"
          onClick={actionsData.btn4Handle}
          loading
        />
      </div>
      <div>
        <BfsButton
          {...args}
          mode="primary"
          label="primary"
          onClick={actionsData.btn1Handle}
          loading
          disabled
        />
        <BfsButton
          {...args}
          mode="secondary"
          label="secondary"
          onClick={actionsData.btn2Handle}
          loading
          disabled
        />
        <BfsButton
          {...args}
          mode="negative"
          label="negative"
          onClick={actionsData.btn3Handle}
          loading
          disabled
        />
        <BfsButton
          {...args}
          mode="positive"
          label="positive"
          onClick={actionsData.btn4Handle}
          loading
          disabled
        />
      </div>
      <div>
        <BfsButton
          {...args}
          mode="primaryGhost"
          label="primary ghost"
          onClick={actionsData.btn5Handle}
        />
        <BfsButton
          {...args}
          mode="secondaryGhost"
          label="secondary ghost"
          onClick={actionsData.btn6Handle}
        />
        <BfsButton
          {...args}
          mode="negativeGhost"
          label="negative ghost"
          onClick={actionsData.btn7Handle}
        />
        <BfsButton
          {...args}
          mode="positiveGhost"
          label="positive ghost"
          onClick={actionsData.btn8Handle}
        />
      </div>
      <div>
        <BfsButton
          {...args}
          mode="primaryGhost"
          label="primary ghost"
          onClick={actionsData.btn1Handle}
          disabled
        />
        <BfsButton
          {...args}
          mode="secondaryGhost"
          label="secondary ghost"
          onClick={actionsData.btn1Handle}
          disabled
        />
        <BfsButton
          {...args}
          mode="negativeGhost"
          label="negative ghost"
          onClick={actionsData.btn1Handle}
          disabled
        />
        <BfsButton
          {...args}
          mode="positiveGhost"
          label="positive ghost"
          onClick={actionsData.btn1Handle}
          disabled
        />
      </div>
      <div>
        <BfsButton
          {...args}
          mode="primaryGhost"
          label="primary ghost"
          onClick={actionsData.btn5Handle}
          loading
        />
        <BfsButton
          {...args}
          mode="secondaryGhost"
          label="secondary ghost"
          onClick={actionsData.btn6Handle}
          loading
        />
        <BfsButton
          {...args}
          mode="negativeGhost"
          label="negative ghost"
          onClick={actionsData.btn7Handle}
          loading
        />
        <BfsButton
          {...args}
          mode="positiveGhost"
          label="positive ghost"
          onClick={actionsData.btn8Handle}
          loading
        />
      </div>
      <div>
        <BfsButton
          {...args}
          mode="primaryGhost"
          label="primary ghost"
          onClick={actionsData.btn1Handle}
          loading
          disabled
        />
        <BfsButton
          {...args}
          mode="secondaryGhost"
          label="secondary ghost"
          onClick={actionsData.btn1Handle}
          loading
          disabled
        />
        <BfsButton
          {...args}
          mode="negativeGhost"
          label="negative ghost"
          onClick={actionsData.btn1Handle}
          loading
          disabled
        />
        <BfsButton
          {...args}
          mode="positiveGhost"
          label="positive ghost"
          onClick={actionsData.btn1Handle}
          loading
          disabled
        />
      </div>
    </div>
  ),
};

export const WithParam = {
  args: {
    label: 'With Params',
    param: {
      key1: 'param 1',
      key2: [],
    },
  },
  render: (args) => <BfsButton {...args} />,
};
