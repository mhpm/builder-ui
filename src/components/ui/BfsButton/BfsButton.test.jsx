import React from 'react';
import userEvent from '@testing-library/user-event'
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { BfsButton } from '.';
import { modes } from './BfsButton';

describe('BfsButton component render', () => {
  it('correctly without crashing, default', () => {
    render(<BfsButton />);
    const button = screen.getByText('button');
    const icon = screen.getByTestId('icon-spinner');

    expect(button).toBeInTheDocument();
    expect(icon.parentElement).toHaveStyle('display: none');
  });

  it('correctly without crashing with tableHeader mode', () => {
    render(<BfsButton data-testid="btn" mode="tableHeader" />);
    const button = screen.getByTestId('btn');
    const icon = screen.getByTestId('icon-spinner');

    expect(button).toBeInTheDocument();
    expect(button).toHaveStyle('display: inline-flex');
    expect(button).not.toHaveStyle('border: solid');
    expect(icon.parentElement).toHaveStyle('display: none');
  });

  it('correctly without crashing with testid', () => {
    render(<BfsButton data-testid="btn" />);
    expect(screen.getByTestId('btn')).toBeInTheDocument();
  });

  it('correctly without crashing with children', () => {
    render(
      <BfsButton data-testid="btn">
        <span data-testid="descendant">btn with children</span>
      </BfsButton>,
    );

    const descendant = screen.getByTestId('descendant');
    expect(screen.getByTestId('btn')).toContainElement(descendant);
  });

  it('correctly without crashing, matching snapshot', () => {
    render(<BfsButton />);
    expect(screen.getByText('button')).toMatchSnapshot();
  });
});

describe('BfsButton component property', () => {
  describe('Label', () => {
    it('set to any text value', () => {
      render(<BfsButton label="save" />);
      expect(screen.getByText('save')).toHaveTextContent(/^save$/);
    });

    it('set to empty value', () => {
      render(<BfsButton data-testid="btn" label="" />);
      expect(screen.getByTestId('btn')).not.toHaveTextContent();
    });
  });

  describe('Loading', () => {
    it('set to true with default icon `spinner`', () => {
      render(<BfsButton data-testid="btn" loading />);

      expect(screen.getByTestId('btn')).toMatchSnapshot();
      expect(screen.getByTestId('icon-spinner')).toBeInTheDocument();
      expect(screen.getByText('button')).toHaveStyle(`opacity: 0`);
    });

    it('set to true with icon `spinner` set', () => {
      render(<BfsButton data-testid="btn" loading icon="spinner" />);

      expect(screen.getByTestId('btn')).toMatchSnapshot();
      expect(screen.getByTestId('icon-spinner')).toBeInTheDocument();
      expect(screen.getByText('button')).toHaveStyle(`opacity: 0`);
    });

    it('set to true with icon `cog` set', () => {
      render(<BfsButton data-testid="btn" loading icon="cog" />);

      expect(screen.getByTestId('btn')).toMatchSnapshot();
      expect(screen.getByTestId('icon-cog')).toBeInTheDocument();
      expect(screen.getByText('button')).toHaveStyle(`opacity: 0`);
    });

    it('set to true with icon `circlenotch` set', () => {
      render(<BfsButton data-testid="btn" loading icon="circlenotch" />);

      expect(screen.getByTestId('btn')).toMatchSnapshot();
      expect(screen.getByTestId('icon-circlenotch')).toBeInTheDocument();
      expect(screen.getByText('button')).toHaveStyle(`opacity: 0`);
    });
  });

  describe('Disable', () => {
    it('set to true', () => {
      render(<BfsButton data-testid="btn" disabled />);
      expect(screen.getByTestId('btn')).toBeDisabled();
    });

    it('set to false', () => {
      render(<BfsButton data-testid="btn" />);
      expect(screen.getByTestId('btn')).toBeEnabled();
    });
  });

  describe('Fluid', () => {
    it('set to true', () => {
      render(<BfsButton data-testid="btn" fluid />);
      expect(screen.getByTestId('btn')).toHaveStyle('width: 100%');
    });

    it('set to false', () => {
      render(<BfsButton data-testid="btn" />);
      expect(screen.getByTestId('btn')).not.toHaveStyle('width: 100%');
    });
  });

  describe('BfsButton mode styles', () => {
    Object.keys(modes).forEach((mode) => {
      it(`applies correct styles for ${mode} mode`, () => {
        render(<BfsButton mode={mode}>Test Button</BfsButton>);
        const button = screen.getByRole('button');

        expect(button).toHaveStyle(`backgroundColor: var(--rol-${mode}-bg-color)`);
      });
    });

    it(`applies correct styles for tableHeader mode`, () => {
      render(<BfsButton mode="tableHeader">Test Button</BfsButton>);
      const button = screen.getByRole('button');

      expect(button).toHaveStyle(`border: none`);
    });
  });
});

describe('BfsButton component interaction', () => {
  describe('mouse', () => {
    it('click with no callback assigned', async () => {
      const user = userEvent.setup();
      const handleClick = jest.fn();

      render(<BfsButton data-testid="btn">Click Me</BfsButton>);
      await user.click(screen.getByTestId('btn'));

      expect(handleClick).not.toHaveBeenCalled();
    });

    it('click', async () => {
      const user = userEvent.setup();
      const handleClick = jest.fn();

      render(<BfsButton data-testid="btn" onClick={handleClick}>Click Me</BfsButton>);
      await user.click(screen.getByTestId('btn'));

      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('click with param', async () => {
      const user = userEvent.setup();
      const handleClick = jest.fn();

      render(<BfsButton data-testid="btn" onClick={handleClick} param="foo">Click Me</BfsButton>);
      await user.click(screen.getByTestId('btn'));

      expect(handleClick).toHaveBeenCalledTimes(1);
      expect(handleClick).toHaveBeenCalledWith('foo');
    });

    it('click with no param', async () => {
      const user = userEvent.setup();
      const handleClick = jest.fn();

      render(<BfsButton data-testid="btn" onClick={handleClick}>Click Me</BfsButton>);
      await user.click(screen.getByTestId('btn'));

      expect(handleClick).toHaveBeenCalledTimes(1);
      expect(handleClick).not.toHaveBeenCalledWith('foo');
    });

    it('click with loading state do not run the callback', async () => {
      const user = userEvent.setup();
      const handleClick = jest.fn();

      render(<BfsButton data-testid="btn" loading onClick={handleClick}>Click Me</BfsButton>);
      await user.click(screen.getByTestId('btn'));

      expect(handleClick).toHaveBeenCalledTimes(0);
    });

    it('click with disable state do not run the callback', async () => {
      const user = userEvent.setup();
      const handleClick = jest.fn();

      render(<BfsButton data-testid="btn" disabled onClick={handleClick}>Click Me</BfsButton>);
      await user.click(screen.getByTestId('btn'));

      expect(handleClick).toHaveBeenCalledTimes(0);
    });
  });

  describe('keyboard', () => {
    it('calls onClick callback when `Escape` key is pressed and escEnable prop is TRUE', async () => {
      const user = userEvent.setup();
      const handleClick = jest.fn();

      render(<BfsButton data-testid="btn" escEnable onClick={handleClick} />);
      const btn = screen.getByTestId('btn');

      await user.tab();
      expect(btn).toHaveFocus();

      await user.keyboard('[Escape/]', btn);
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('calls onClick callback when `Escape` key is pressed and escEnable prop is FALSE', async () => {
      const user = userEvent.setup();
      const handleClick = jest.fn();

      render(<BfsButton data-testid="btn" onClick={handleClick} />);
      const btn = screen.getByTestId('btn');

      await user.tab();
      expect(btn).toHaveFocus();

      await user.keyboard('[Escape/]', btn);
      expect(handleClick).toHaveBeenCalledTimes(0);
    });

    it('calls onClick callback when `Space` key is pressed', async () => {
      const user = userEvent.setup();
      const handleClick = jest.fn();

      render(<BfsButton data-testid="btn" onClick={handleClick} />);
      const btn = screen.getByTestId('btn');

      await user.tab();
      expect(btn).toHaveFocus();

      await user.keyboard('[Space/]', btn);
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('calls onClick callback when `Enter` key is pressed', async () => {
      const user = userEvent.setup();
      const handleClick = jest.fn();

      render(<BfsButton data-testid="btn" onClick={handleClick} />);
      const btn = screen.getByTestId('btn');

      await user.tab();
      expect(btn).toHaveFocus();

      await user.keyboard('[Enter/]', btn);
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('calls onClick callback when any other key is pressed', async () => {
      const user = userEvent.setup();
      const handleClick = jest.fn();

      render(<BfsButton data-testid="btn" onClick={handleClick} />);
      const btn = screen.getByTestId('btn');

      await user.tab();
      expect(btn).toHaveFocus();

      await user.keyboard('[a/]', btn);
      expect(handleClick).toHaveBeenCalledTimes(0);
    });
  });
});
