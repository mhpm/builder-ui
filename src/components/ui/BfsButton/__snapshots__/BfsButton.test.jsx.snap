// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BfsButton component render Correctly without crashing, matching snapshot 1`] = `
<span
  class="content"
>
  button
</span>
`;
